processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java

```diff
@@ -96,27 +144,22 @@ public boolean equals(Object o)
 
     SelectorDimFilter that = (SelectorDimFilter) o;
 
-    if (dimension != null ? !dimension.equals(that.dimension) : that.dimension != null) {
+    if (!dimension.equals(that.dimension)) {
       return false;
     }
     if (value != null ? !value.equals(that.value) : that.value != null) {
       return false;
     }
+    return extractionFn != null ? extractionFn.equals(that.extractionFn) : that.extractionFn == null;
```

Comment:
@drcrallen Only BoundDimFilter and SelectorDimFilter have proper equals implementations;

I had the equals() and hashcode() methods for those DimFilters throw an exception when called, and found out that AndDimFilter and OrDimFilter trigger equals() on its sub filters via this call:
```
    fields.removeAll(Collections.singletonList(null));
```

But the default equals() implementation is sufficient for this null removal.

I also disabled the proper equals/hashcode methods on BoundDimFilter and SelectorDimFilter, the only problem I saw with doing that in the unit tests was in TimeseriesQueryRunnerTest, where there are some explicit equality checks on SelectorDimFilter:
```
    Assert.assertEquals(
        Druids.newSelectorDimFilterBuilder()
              .dimension(QueryRunnerTestHelper.marketDimension)
              .value("upfront")
              .build(),
        query.getDimensionsFilter()
    );
```